builder/azure/chroot/diskattacher.go

```diff
@@ -8,15 +8,16 @@ import (
 	"time"
 
 	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-12-01/compute"
-	"github.com/Azure/go-autorest/autorest/azure"
 	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/hashicorp/packer-plugin-azure/builder/azure/common/client"
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
 )
 
 type DiskAttacher interface {
-	AttachDisk(ctx context.Context, disk string) (lun int32, err error)
+	AttachDisk(ctx context.Context, state multistep.StateBag, disk string) (lun int32, err error)
```

Comment:
As opposed to passing in the entire statebag to get to the UI. I would recommend adding a field to each of these types for setting the UI. This way you can access the UI simply by calling `da.ui.(....)`. 

For example given the DiskAttacher type you could add the field
```go
type diskAttacher struct {
	azcli client.AzureClientSet
	vm *client.ComputeInfo // store info about this VM so that we don't have to ask metadata service on every call
        ui packersdk.Ui
}

var NewDiskAttacher = func(azureClient client.AzureClientSet, ui packersdk.Ui) DiskAttacher {
	return &diskAttacher{
		azcli: azureClient,
                ui: ui
	}
}

func (da *diskAttacher) DetachDisk(ctx context.Context, diskID string) error {
 //...

	log.Println("Fetching list of disks currently attached to VM")
	currentDisks, err := da.getDisks(ctx)
	if err != nil {
....
		da.ui.Say("Initial call for fetching VM instance disks returned an error. Checking to see if instance is part of a VM scale set before giving up.")

```
Then in the using the unit test as an example you could write
```go

	testDiskName := t.Name()

	errorBuffer := &strings.Builder{}
	ui := &packersdk.BasicUi{
		Reader:      strings.NewReader(""),
		Writer:      ioutil.Discard,
		ErrorWriter: errorBuffer,
	}

	state := new(multistep.BasicStateBag)
	state.Put("azureclient", azcli)
        da := NewDiskAttacher(azcli, ui)
```

TBH it feels a little weird to me to have all these exported fields and would prefer to have Ui and AZcli as the field names. But I think sticking with the style that exists is fine for now.

Reply:
Fully agree with you - passing the state felt a bit too much. Thanks for the suggestion. I managed to incorporate the new field in my latest commit ([395277a](https://github.com/hashicorp/packer-plugin-azure/pull/184/commits/395277ae7283f2930fa68eb4a94fa06ee8da32c9)). It should all be fine now.